(Multiple) Local Playlist (instead of saving video in one hidden list)

PikachuEXE opened the pull request:
## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [x] Feature Implementation

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->
- No idea please update this list directly lol

## Description
<!-- Please write a clear and concise description of what the pull request does. -->
Instead of saving videos (in one hidden playlist) and displaying only that list
This PR makes it possible to manage multiple local playlists and adding videos to it
You can also import export

## Tasks Left
- [x] Translation (already done but could have some entries missed)
- [ ] Code Cleanup

## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->
Lazy maybe post later on request

## Testing <!-- for code that is not small enough to be easily understandable -->
<!-- Has this pull request been tested? -->
<!-- Please describe shortly how you tested it. -->
<!-- Are there any ramifications remaining? -->
- Playlist migration (from old version of FT)
- Playlist import export (from old/new version of FT)
- Playlist management
- Video management in playlist (including adding video(s) to playlist)
- I18n (not sure how to test this)
- What else?

## Additional context
<!-- Add any other context about the pull request here. -->
Code based on branch https://github.com/FreeTubeApp/FreeTube/tree/playlists
See https://github.com/PikachuEXE/FreeTube/issues for issues reported before this PR created

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Note: See playlist related issues https://github.com/PikachuEXE/FreeTube/issues/44

Ill check later in that list if this PR closes more issues

efb4f5ff-1298-471a-8973-3d47447115dc commented:
It should close #2728 but i would like the others to confirm this

ChunkyProgrammer commented on src/renderer/components/watch-video-info/watch-video-info.js:
```diff
@@ -268,15 +256,32 @@ export default defineComponent({
+        playlistLoop: this.getPlaylistLoop(),
+      }
+      // Only play video in non playlist mode when user playlist detected
+      if (!this.inUserPlaylist) {
```
Shouldn't this be
```suggestion
      if (this.inUserPlaylist) {
```

PikachuEXE replied:
Fixed

jasonhenriquez commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,85 @@
+    this.$refs.playlistNameInput.focus()
+  },
+  destroyed() {
+    if (this.lastActiveElement != null) {
+      this.lastActiveElement.focus()
+    }
+  },
```
```suggestion
    this.lastActiveElement?.focus()
  },
```

efb4f5ff-1298-471a-8973-3d47447115dc replied:
I think you marked this as resolved on accident

jasonhenriquez commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,85 @@
+    this.$refs.playlistNameInput.focus()
+  },
+  destroyed() {
+    if (this.lastActiveElement != null) {
+      this.lastActiveElement.focus()
+    }
```
```suggestion
   this.lastActiveElement?.focus()
```

jasonhenriquez commented on src/renderer/components/ft-prompt/ft-prompt.js:
```diff
@@ -64,6 +67,11 @@ export default defineComponent({
     this.focusItem(0)
   },
+  destroyed() {
+    if (this.lastActiveElement != null) {
+      this.lastActiveElement.focus()
+    }
```
```suggestion
   this.lastActiveElement?.focus()
```

jasonhenriquez replied:
Lots of examples of this, so just do a find and replace

PikachuEXE replied:
I only see 3 (search by `Element.focus()`)

jasonhenriquez commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.js:
```diff
@@ -0,0 +1,267 @@
+    // User might want to search first if they have many playlists
+    this.$refs.searchBar.focus()
+  },
+  destroyed() {
+    if (this.lastActiveElement != null) {
+      this.lastActiveElement.focus()
```
```suggestion
   this.lastActiveElement?.focus()
```

absidue commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,83 @@
+import Vue from 'vue'
+import { mapActions } from 'vuex'
+import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
+import FtPrompt from '../ft-prompt/ft-prompt.vue'
+import FtButton from '../ft-button/ft-button.vue'
+import FtInput from '../ft-input/ft-input.vue'
+import FtPlaylistSelector from '../ft-playlist-selector/ft-playlist-selector.vue'
+import {
+  showToast,
+} from '../../helpers/utils'
+
+export default Vue.extend({
```
`defineComponent` gives us auto-complete and is compatible with Vue 3.
```suggestion
import { defineComponent } from 'vue'
import { mapActions } from 'vuex'
import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
import FtPrompt from '../ft-prompt/ft-prompt.vue'
import FtButton from '../ft-button/ft-button.vue'
import FtInput from '../ft-input/ft-input.vue'
import FtPlaylistSelector from '../ft-playlist-selector/ft-playlist-selector.vue'
import {
  showToast,
} from '../../helpers/utils'

export default defineComponent({
```

absidue commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.js:
```diff
@@ -0,0 +1,75 @@
+import Vue from 'vue'
+import FtIconButton from '../ft-icon-button/ft-icon-button.vue'
+import { mapActions } from 'vuex'
+
+export default Vue.extend({
```
`defineComponent` gives us auto-complete and is compatible with Vue 3.
```suggestion
import { defineComponent } from 'vue'
import FtIconButton from '../ft-icon-button/ft-icon-button.vue'
import { mapActions } from 'vuex'

export default defineComponent({
```

absidue commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.js:
```diff
@@ -0,0 +1,265 @@
+import Vue from 'vue'
+import { mapActions } from 'vuex'
+import debounce from 'lodash.debounce'
+import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
+import FtPrompt from '../ft-prompt/ft-prompt.vue'
+import FtButton from '../ft-button/ft-button.vue'
+import FtPlaylistSelector from '../ft-playlist-selector/ft-playlist-selector.vue'
+import FtInput from '../../components/ft-input/ft-input.vue'
+import FtSelect from '../../components/ft-select/ft-select.vue'
+import {
+  showToast,
+} from '../../helpers/utils'
+
+const SORT_BY_VALUES = {
+  NameAscending: 'name_ascending',
+  NameDescending: 'name_descending',
+
+  LatestCreatedFirst: 'latest_created_first',
+  EarliestCreatedFirst: 'earliest_created_first',
+
+  LatestUpdatedFirst: 'latest_updated_first',
+  EarliestUpdatedFirst: 'earliest_updated_first',
+}
+
+export default Vue.extend({
```
`defineComponent` gives us auto-complete and is compatible with Vue 3.
```suggestion
import { defineComponent } from 'vue'
import { mapActions } from 'vuex'
import debounce from 'lodash.debounce'
import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
import FtPrompt from '../ft-prompt/ft-prompt.vue'
import FtButton from '../ft-button/ft-button.vue'
import FtPlaylistSelector from '../ft-playlist-selector/ft-playlist-selector.vue'
import FtInput from '../../components/ft-input/ft-input.vue'
import FtSelect from '../../components/ft-select/ft-select.vue'
import {
  showToast,
} from '../../helpers/utils'

const SORT_BY_VALUES = {
  NameAscending: 'name_ascending',
  NameDescending: 'name_descending',

  LatestCreatedFirst: 'latest_created_first',
  EarliestCreatedFirst: 'earliest_created_first',

  LatestUpdatedFirst: 'latest_updated_first',
  EarliestUpdatedFirst: 'earliest_updated_first',
}

export default defineComponent({
```

absidue commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.scss:
This file doesn't contain any scss specific syntax, so it can be renamed to a `.css` file, please remember to change it in the template (`.vue`) file too.

absidue commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.vue:
```diff
@@ -0,0 +1,70 @@
+      <ft-flex-box>
+        <ft-playlist-selector
+          v-for="(playlist, index) in activePlaylists"
+          :key="`${playlist._id}-${index}`"
```
If playlists have unique `_id`s, you can remove the index from the key.

absidue commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.js:
```diff
@@ -0,0 +1,75 @@
+      return `${this.title.substring(0, 255)}...`
+    },
+  },
+  mounted: function () {
+    this.parseUserData()
+  },
```
```suggestion
  created: function () {
    this.parseUserData()
  },
```

absidue commented on src/renderer/components/ft-prompt/ft-prompt.js:
```diff
@@ -64,6 +67,9 @@ export default defineComponent({
     })
     this.focusItem(0)
   },
+  destroyed() {
+    this.lastActiveElement?.focus()
+  },
```
`destroyed` is called after the component instance has been destroyed, so `this` is no longer available at that point. `beforeDestroy` is called just before the component instance is destroyed, so `this` is still available.
```suggestion
  beforeDestroy() {
    this.lastActiveElement?.focus()
  },
```

absidue commented on src/renderer/components/ft-share-button/ft-share-button.js:
```diff
@@ -62,6 +62,19 @@ export default defineComponent({
     },
 
+    selectedUserPlaylist: function () {
+      if (this.playlistId == null) { return null }
+      if (this.playlistId === '') { return null }
```
These two `if`s can be combined into a single `if`.
```suggestion
      if (this.playlistId == null || this.playlistId === '') { return null }
```

absidue commented on src/renderer/components/watch-video-info/watch-video-info.js:
```diff
@@ -268,15 +256,32 @@ export default defineComponent({
+      this.openInExternalPlayer(payload)
+    },
+
+    handleFormatChange: function (format) {
+      switch (format) {
+        case 'dash':
+          this.$parent.enableDashFormat()
+          break
+        case 'legacy':
+          this.$parent.enableLegacyFormat()
+          break
+        case 'audio':
+          this.$parent.enableAudioFormat()
+          break
```
This function was removed a while ago (the template uses `$emit`, instead of calling those functions with `$parent`), I'm guessing this was because development was merged into here, instead of this branch being rebased on top of development.

absidue commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -138,8 +219,6 @@ export default defineComponent({
-      this.isLoading = true
-
       invidiousGetPlaylistInfo(this.playlistId).then((result) => {
         this.infoData = {
```
`infoData` doesn't seem to be used anymore, so any code relating to it should probably be removed.

absidue commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -59,20 +72,78 @@ export default defineComponent({
+      return this.$store.getters.getPlaylistsReady
+    },
+    selectedUserPlaylist: function () {
+      if (this.playlistId == null) { return null }
+      if (this.playlistId === '') { return null }
```
Can be combined to a single `if`.
```suggestion
      if (this.playlistId == null || this.playlistId === '') { return null }
```

absidue reviewed (commented):
Haven't reviewed everything yet but here is an initial code review.

Some of the things I noticed that I didn't leave individual comments on, is that especially in the CSS and SCSS files, the newly added lines don't have any whitespace between them, which makes them hard to read.

Also please go through your changes and make sure you aren't reintroducing code that was removed ages ago, I noticed one case in `watch-video-info` but there might be more.

absidue commented on src/renderer/components/playlist-info/playlist-info.js:
```diff
@@ -79,49 +164,144 @@ export default defineComponent({
+      return !this.hideSharingActions
+    },
+  },
+  mounted: function () {
+    this.newTitle = this.title
+    this.newDescription = this.description
```
```suggestion
  created: function () {
    this.newTitle = this.title
    this.newDescription = this.description
```

PikachuEXE commented:
Changes are made
The new files are all from the existing old playlist branch so they contain old code
Please review them harder

jasonhenriquez commented on src/renderer/store/modules/playlists.js:
```diff
@@ -1,32 +1,83 @@
-      videos: []
-    }
-  ]
+      playlistName: 'Watch Later',
```
Discussion on this: It seems like the current Favorites functionality in FreeTube today is the equivalent of a Watch Later. Why is this being separated in this manner? More broadly, I want to pick your brain on what your plans for this are. My top concerns for this PR are:
1. Maintaining reverse compatibility for users who download the new version but want to revert for whatever reason
2. Adding back all of the removed instances of the favorite icon in the JS, template, and CSS files with a functionally equivalent or better Add to Watch Later button. I think having a release without this present would be a strong negative change, as our dedicated users are the ones who have adapted to saving videos in this one-click way we have set up. Because of the sheer importance of this, I want to make sure that your plans for this are made clear for both the sake of discussion and for allowing other devs to follow through with this in a way that's aligned with your vision if you were unable to complete development on this feature.

ChunkyProgrammer replied:
@PikachuEXE thoughts on these two points ?(in case you missed the notification)

PikachuEXE replied:
> It seems like the current Favorites functionality in FreeTube today is the equivalent of a Watch Later.

It might be used that way (or not, since I don't use it as watch later ever), since there is **only one playlist** before this PR.

1. Reverting is gonna be moving (copying is fine) all videos into **the first playlist** (well current version of FT just display the list of videos of first playlist, you can check exported playlist files). It's not automatic but if you want a version which has automatic revert function we will need to release a version just for that (before releasing a version with this PR)
2. I would NOT add more functionality to this PR (which is already a complicated one). The current UX is based on having **one playlist** only and I would not consider that is good for UX with privacy (over one-click features). Feel free to develop and propose another PR if anyone is willing to.

jasonhenriquez commented:
Not sure if this is pertinent, but changing from this branch to `development` is the first time I had to delete `dist` and `node_modules` in order to prevent `npm run dev` from failing with an error. It was persisting as an issue even after switching back to this branch again and clearing my `Electron` folder.

absidue commented:
When you go to add a video to a playlist, in the prompt hovering over the playlist shows the normal cursor instead of the pointer/hand, which you would get when you hover over a button or a link, it would be nice if the cursor changed, so it's clear that you can click on the playlists.

PikachuEXE commented on src/renderer/store/modules/playlists.js:
```diff
@@ -64,13 +170,118 @@ const actions = {
+        })
+        const findWatchLater = payload.filter((playlist) => {
+          return playlist.playlistName === 'Watch Later' || playlist._id === 'watchLater'
+        })
```
Fixed
I blame old code :)

PikachuEXE commented on src/renderer/store/modules/playlists.js:
```diff
@@ -1,32 +1,83 @@
+
+function generateRandomUniqueId() {
+  return crypto.randomUUID()
+}
```
Fixed

PikachuEXE commented on src/renderer/scss-partials/_ft-list-item.scss:
```diff
@@ -148,11 +148,29 @@ $watched-transition-duration: 0.5s;
+    .downArrowIcon {
+      font-size: 17px;
       block-size: fit-content;
     }
```
If there are class specific style later I would re-expand this
For now it's ok to be in one

<details>

<summary>I saw style declared somewhere else like this:</summary>

```scss
.a, .b, .c {
  // common style
}

.a {
  // specific style
}
.b {
  // specific style
}
.c {
  // specific style
}
```

What I want:
```scss

.a {
  // common + specific style
}
.b {
  // common + specific style
}
.c {
  // common + specific style
}
```

</details>

PikachuEXE commented on src/renderer/components/watch-video-playlist/watch-video-playlist.vue:
```diff
@@ -114,13 +118,13 @@
           v-for="(item, index) in playlistItems"
-          :key="index"
-          :ref="currentVideoIndex === (index + 1) ? 'currentVideoItem' : null"
+          :key="`index-${item.playlistItemId || item.videoId}`"
```
Fixed
Totally forgot why there is a prefix

PikachuEXE commented on src/renderer/components/watch-video-playlist/watch-video-playlist.vue:
```diff
@@ -17,27 +17,31 @@
       <span
-        v-else
-        class="channelName"
+        v-if="channelName !== ''"
```
Fixed and TIL (maybe just forgot)

ChunkyProgrammer commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,83 @@
+        return playlist.playlistName === this.playlistName
+      })
+      if (nameExists !== -1) {
+        showToast('There is already a playlist with this name. Please pick a different name.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,83 @@
+  methods: {
+    createNewPlaylist: function () {
+      if (this.playlistName === '') {
+        showToast('Playlist name cannot be empty. Please input a name.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,83 @@
+
+      try {
+        this.addPlaylist(playlistObject)
+        showToast(`Playlist ${this.playlistName} has been successfully created.`)
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.js:
```diff
@@ -0,0 +1,83 @@
+        showToast(`Playlist ${this.playlistName} has been successfully created.`)
+      } catch (e) {
+        showToast('There was an issue with creating the playlist.')
+        console.error(e)
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.js:
```diff
@@ -0,0 +1,265 @@
+        this.updatePlaylist({ _id: playlist._id })
+      })
+
+      showToast(`${this.toBeAddedToPlaylistVideoCount} video(s) added to ${addedPlaylistIds.size} playlist(s).`)
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/components/privacy-settings/privacy-settings.js:
```diff
@@ -114,6 +115,14 @@ export default defineComponent({
+      if (option !== 'yes') { return }
+
+      this.removeAllPlaylists()
+      showToast('All playlists have been removed')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+      })
+
+      if (videoIndex === 0) {
+        showToast('This video cannot be moved up.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+        this.updatePlaylist(playlist)
+        this.playlistItems = playlistItems
+      } catch (e) {
+        showToast('There was an issue with updating this playlist.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+      })
+
+      if (videoIndex + 1 === playlistItems.length || videoIndex + 1 > playlistItems.length) {
+        showToast('This video cannot be moved down.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+        this.updatePlaylist(playlist)
+        this.playlistItems = playlistItems
+      } catch (e) {
+        showToast('There was an issue with updating this playlist.')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+        this.updatePlaylist({ _id: this.playlistId })
+        showToast('Video has been removed')
+      } catch (e) {
+        showToast('There was a problem with removing this video')
```
Toast message needs localization

ChunkyProgrammer commented on src/renderer/views/Playlist/Playlist.js:
```diff
@@ -210,8 +295,90 @@ export default defineComponent({
+        })
+        // Update playlist's `lastUpdatedAt`
+        this.updatePlaylist({ _id: this.playlistId })
+        showToast('Video has been removed')
```
Toast message needs localization

PikachuEXE commented:
@ChunkyProgrammer 
All toast messages (playlist related) should be translated now

Also fixed old code for "remove watched videos" which was broken by https://github.com/FreeTubeApp/FreeTube/pull/4017

absidue commented:
Yep, when you maintain a separate branch for that long, you need to stay on top of all changes that are made to the original branch and replicate them in your branch.

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
This is kinda unfortunate


https://github.com/FreeTubeApp/FreeTube/assets/73130443/dafea57e-7f8f-4834-b3ac-04de2dde4a08

absidue commented:
Blame YouTube, we don't know that the continuation is going to be empty until we request it and get an empty response.

PikachuEXE commented:
That's related to YT returning fake continuation token

https://github.com/FreeTubeApp/FreeTube/pull/4102

But to fix the remote playlist issue you mentioned here requires eager loading all items in remote playlist which is probably not what we want (and it's also not caused by this PR, checked `development`)

ChunkyProgrammer commented on src/renderer/views/UserPlaylists/UserPlaylists.vue:
```diff
@@ -1,47 +1,61 @@
+          {{ $t("User Playlists.Your Playlists") }}
+        </h2>
+        <ft-icon-button
+          title="Create New Playlist"
```
This should be translated

ChunkyProgrammer commented on src/renderer/components/playlist-info/playlist-info.vue:
```diff
@@ -68,11 +98,71 @@
+      />
+      <ft-prompt
+        v-if="showRemoveVideosOnWatchPrompt"
+        :label="$t('User Playlists.Are you sure you want to remove all watched videos from this playlist? This cannot be undone.')"
```
the . at the end is causing issues when called like this. You can either remove the . from the key or do the following change
```suggestion
        :label="$t('User Playlists[\'Are you sure you want to remove all watched videos from this playlist? This cannot be undone.\']')"
```

ChunkyProgrammer commented on src/renderer/components/playlist-info/playlist-info.vue:
```diff
@@ -68,11 +98,71 @@
+
+      <ft-prompt
+        v-if="showDeletePlaylistPrompt"
+        :label="$t('User Playlists.Are you sure you want to delete this playlist? This cannot be undone.')"
```
same issue with `.` at the end

PikachuEXE replied:
Fixed, hard to spot thx

ChunkyProgrammer commented on src/renderer/components/data-settings/data-settings.js:
```diff
@@ -882,87 +882,116 @@ export default defineComponent({
-        'isLive',
-        'type',
+
+        // `playlistItemId` should be optional for backward compatibility
```
Shouldn't this be added to optionalKeys then? Or is there a reason it wasn't added?

PikachuEXE replied:
If you look at the video import code, the whole video object is pushed as long as all the required keys are present (not whitelisting properties)

ChunkyProgrammer commented on src/renderer/store/modules/playlists.js:
```diff
@@ -1,32 +1,83 @@
-      videos: []
-    }
-  ]
+      playlistName: 'Watch Later',
```
Renaming 'Watch Later' to 'WatchLater' imported an empty playlist when importing an old version of FreeTube's playlists
![image](https://github.com/FreeTubeApp/FreeTube/assets/78101139/040c4455-65d2-42eb-99d2-8b0bf4a0e62f)

I also wonder if we should keep these default playlists protected. IMO it could be a good idea to keep Watch Later and Favorites as protected playlists, then we can also add translations to them

PikachuEXE replied:
You can reintroduce "default special playlists" later
Right now if you have protected, that means more functions & special playlist treatment need to be added and tested (which might still look incomplete due to playlist translation?)
Having an empty `WatchLater` (no space) playlist is an one time issue and I don't want to exclude that playlist (again that would be special playlist treatment)

PikachuEXE replied:
Also the idea of supporting multiple playlists is to have multiple "favorites" & "watch later" playlists (however they want to use their created playlist, as long term or temporary bookmark is their thing, we just need to enable them to do both). Until you can come up with features linked to "special playlists", those playlists shouldn't even be introduced, not to say protected.

jasonhenriquez replied:
The main existent feature linked to a special playlist for Favorites is maintaining reverse compatibility regardless of user actions, no? That, and signalling to the user that those playlists do possess some special importance (reverse compatibility now, Watch Later / Favorites shortcut down the road). + Allows us to control the (translated) labels for these playlists at any point down the road for all users.

PikachuEXE replied:
Keeping non special playlists protected instead of allowing user to opt-in and choose the playlist(s) for later introduced features like "quick save video for purpose of long term bookmark and/or watch later"?

jasonhenriquez replied:
I personally think customizing which playlist(s) the Watch Later button is for is a more debatable and lower priority feature than having one fixed & designated one. In any case, having one playlist that occupies the special position of representing legacy Favorites and one that occupies the special position of Watch Later (note: I'm fine with these two being the same) that are both protected is preferable. It provides a sense of constancy to the user and emphasizes their specialness for both reverse compatibility and Watch Later. The implementation details of the power user benefit of being able to delete and rename them is IMO not that important and can be revisited for when we do implement customizing which playlist(s) the Watch Later button is for.

PikachuEXE replied:
If anyone have developed/is developing the quick save video feature, I am fine it's not selectable at the beginning but the special playlist should be able to be hidden (when the feature is disabled).

Reverse compatibility can be done IMO via either by instructions in document (move all videos to one playlist) or by code (not sure how but that means another release required before this is merged).

Either way there is still no place for special playlists in this PR.

jasonhenriquez commented:
Sorry about that; merged into the wrong branch a short time back & had some issues with fixing the history. Should be fine now, even if ugly. Not been having the best week. 🤦

PikachuEXE commented:
Force pushed with new commits from my local to fix it

jasonhenriquez reviewed (commented):
**issue (unsure if blocking):** when reverting to the development branch, the playlist feature is broken (can't remove any of my favorites). This is going to be a notable issue for the dozens or hundreds of people who choose to load up the new version and revert to an earlier one at some point. They are likely going to be very confused about this. Do we have the upgrade / downgrade issues catalogued in any created issue? Because I worry this could be a big problem.

Our options are to try to fix this, if possible, by making the conversion process more reverse compatible, or add documentation in the release notes and/or in the code itself notifying users that this will occur, and what they need to do to fix it (create a backup of their settings before/after (?), then delete your `settings.db` (?)). But this might just be a blocking issue.

jasonhenriquez commented:
**issue (non-blocking, minor):** sort order broken once search is in use on the `ft-playlist-add-video-prompt`

![Screenshot_20231118_215226](https://github.com/FreeTubeApp/FreeTube/assets/84899178/d7ba8434-4f19-456b-bc8f-b873f862f086)

jasonhenriquez commented on src/renderer/scss-partials/_ft-list-item.scss:
```diff
@@ -148,11 +148,20 @@ $watched-transition-duration: 0.5s;
+
+      display: grid;
+      grid-auto-flow: column;
+      justify-content: flex-end;
+    }
+    .addToPlaylistIcon,
+    .trashIcon,
+    .upArrowIcon,
+    .downArrowIcon {
+      font-size: 17px;
       block-size: fit-content;
```
```suggestion
      justify-content: flex-end;
      block-size: fit-content;
    }
    .addToPlaylistIcon,
    .trashIcon,
    .upArrowIcon,
    .downArrowIcon {
      font-size: 17px;
```
This bug was re-introduced. Here's the fix.

![Screenshot_20231118_220012](https://github.com/FreeTubeApp/FreeTube/assets/84899178/2f5bd36c-1016-4b4e-b5fa-a7e0f8abea4a)

jasonhenriquez replied:
(Test in your own branch first - I tend to make minor linting mistakes on the Web UI - and you can get rid of the `block-size: fit-content;` on line 165 as well)

PikachuEXE replied:
fixed

jasonhenriquez commented on src/renderer/components/playlist-info/playlist-info.vue:
```diff
@@ -19,19 +22,46 @@
+        </span>
+        <span>- </span>
         <span v-if="infoSource !== 'local'">
           {{ $t("Playlist.Last Updated On") }}
```
**nitpick:** this text should be lowercase: `0 Videos - Last Updated On Nov 18, 2023` -> `0 videos - last updated on Nov 18, 2023`. Looks odd in all-caps & out of place with our other similar labels.

PikachuEXE replied:
Please perform case related text update for **existing** translation entries in a separate PR

jasonhenriquez commented on src/renderer/components/playlist-info/playlist-info.vue:
```diff
@@ -68,11 +98,71 @@
+        <ft-icon-button
+          v-if="!editMode && infoSource === 'user' && videoCount > 0"
+          :title="$t('User Playlists.Remove Watched Videos')"
+          :icon="['fas', 'eye-slash']"
```
**issue (non-blocking):** The purpose of this button is incredibly unclear by looking at the icon. This should be less of an issue once we have icon labels, but please see if you can find a better one.

**issue (non-blocking):** On top of the previous issue, this is a destructive action, which we should have clearly marked. In the design world, destructive controls are colored red to clearly communicate to the user that they have a lasting destructive impact. The prompt will hopefully stop most people from making a mistake, but it's still a notable issue. My suggestion is to Implement a color override property on the `ft-icon-button` component and use enforce a red color here (e.g., the red primary color variable and the primary-text-color for that).

PikachuEXE replied:
https://fontawesome.com/search?q=slash&o=r&m=free 
Not that many icons have slash, help me find a better icon I guess ._.

PikachuEXE replied:
Color is updated, I kind of waiting for any of you to comment on it

jasonhenriquez replied:
Yeah, I don't see anything else better in the `fontawesome` list by itself ([without using layers, at least](https://origin.fontawesome.com/docs/web/style/layer)), so that's fine.

jasonhenriquez replied:
The color is updated to use the primary color, not red, which is different than what I was intending. Main point is that forcing red in particular endows it with special meaning (for everyone else except someone with a `Secondary Color Theme` of `Red`, I guess, which you can mitigate by using a darker red.

PikachuEXE replied:
I am following https://github.com/FreeTubeApp/FreeTube/blob/development/src/renderer/components/privacy-settings/privacy-settings.vue#L43-L62
I think it's better to have later PRs to do this kind of minor changes which probably also affects other places

jasonhenriquez replied:
Sounds good, I've placed this into its own issue

jasonhenriquez commented on src/renderer/views/Playlist/Playlist.scss:
```diff
@@ -30,6 +33,30 @@
+}
+.playlistItem-enter, .playlistItem-leave-to {
+  opacity: 0;
+  transform: translateX(10%);
```
```suggestion
  translateX(calc(10% * var(--horizontal-directionality-coefficient))
```

(disclaimer: validate in local before trying; I sometimes make mistakes typing in the GH UI)

PikachuEXE replied:
Seems missing `transform: ` prefix, and semicolon at the end...
Anyway I updated and validated it locally but you should check again

jasonhenriquez replied:
Thank you lol. Tested by adding/removing in English and Arabic. Looks good now!

jasonhenriquez commented on static/locales/en-US.yaml:
```diff
@@ -137,8 +137,78 @@ User Playlists:
     videos currently here will be migrated to a 'Favorites' playlist.
   Your saved videos are empty. Click on the save button on the corner of a video to have it listed here: Your saved videos are empty. Click on the save button on the corner of a video to have
     it listed here
+  You have no playlist. Click on the create new playlist button to create a new one.: You have no playlist. Click on the create new playlist button to create a new one.
```
```suggestion
  You have no playlists. Click on the create new playlist button to create a new one.: You have no playlists. Click on the create new playlist button to create a new one.
```

PikachuEXE replied:
fixed

jasonhenriquez commented on static/locales/en-US.yaml:
```diff
@@ -331,6 +401,9 @@ Settings:
       undone.
+    Remove All Playlists: Remove All Playlists
+    All playlists have been removed: All playlists have been removed
+    Are you sure you want to remove all your playlists?: Are you sure you want to remove all your playlists?
```
```suggestion
    Are you sure you want to remove all of your playlists?: Are you sure you want to remove all of your playlists?
```

PikachuEXE replied:
Any reason to use `all of`?
If it feels inconsistent with `Are you sure you want to remove all subscriptions and profiles?` I can remove `your` (not sure why I add `your`)

jasonhenriquez replied:
I personally think it sounds more formal, but after checking online, [English is weird and both are completely fine](https://english.stackexchange.com/questions/80412/is-of-necessary-in-all-of). Fine to close this one if you want.

jasonhenriquez commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.js:
```diff
@@ -0,0 +1,268 @@
+
+    createPlaylistPromptShown(value) {
+      // Lock interactions when create playlist prompt opened
+      this.interactionsLocked = value
```
**question:** For what purpose are we locking the search bar? Could you document it in the code?

jasonhenriquez replied:
Also, it seems to be just making it non-keyboard-focusable, not disabling it. If you want to disable it, you can use the `disabled` prop on `ft-input`.

jasonhenriquez replied:
**suggestion (blocking):** If you do end up needing to keep this in as is, please replace the `inputTabindex` `number` prop with a `boolean` `isInputKeyboardNavigable`. That way it's less declarative and has no need of handling odd inputs.

PikachuEXE replied:
Code removed due to the issue I experienced at early stage of dev seems no longer reproducible

jasonhenriquez replied:
Thanks! Can you also remove the `input-tabindex` prop from `ft-input`s?

PikachuEXE replied:
Just did

PikachuEXE commented:
> sort order broken once search is in use on the ft-playlist-add-video-prompt

Fixed, got old code for sorting not removed (which was only run for filtered result only)

PikachuEXE commented:
For "can't remove any of my favorites", please see https://github.com/FreeTubeApp/FreeTube/pull/4353

PikachuEXE commented:
For code to allow user to export playlist for older versions of FT, see
https://github.com/PikachuEXE/FreeTube/tree/feature/playlist-2023-05-w-old-ft-consumable-playlist-export

### Testing
- Checkout this PR, add some playlists and videos, probably also removed the default playlists
- Export playlists as usual (also as backup file for different tests
- Export playlists with new toggle on
- Remove all playlists
- Switch back to `development`, import the export with new toggle on
- Confirm all videos from all playlists are present

![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/5b173a90-f457-4006-9690-33e2292991a4)

I prefer not to put it into this PR (already a mega thread...)

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Found errors:

https://github.com/FreeTubeApp/FreeTube/assets/73130443/1bbe0137-3dc8-45e8-b753-7f32c52bc569

[Should we mark these destructive buttons red?](https://github.com/FreeTubeApp/FreeTube/issues/1815)


https://github.com/FreeTubeApp/FreeTube/assets/73130443/ccabc2ee-38c5-4571-9d88-2f5f4af8ab19

PikachuEXE commented:
Vue warning fixed

For button color please see my previous reply in https://github.com/FreeTubeApp/FreeTube/pull/4234#discussion_r1398640085

jasonhenriquez commented on src/renderer/components/ft-button/ft-button.js:
```diff
@@ -18,6 +18,11 @@ export default defineComponent({
       type: String,
       default: ''
-    }
+    },
```
**question:** Why was this added?

jasonhenriquez commented on src/renderer/components/ft-button/ft-button.js:
```diff
@@ -18,6 +18,11 @@ export default defineComponent({
       type: String,
       default: ''
-    }
+    },
+    tabindex: {
+      type: Number,
+      // Relies on template to not output tabindex on values < -1
+      default: -2,
+    },
```
**question:** Why was this added?

PikachuEXE replied:
Probably part of the workaround I introduced before
Let me see if they can be removed

PikachuEXE replied:
Removed
Probably wanted to set as `-1` at some point as workaround to some issue which can no longer be reproduced
Forgot what issue

PikachuEXE commented:
Just merged https://github.com/PikachuEXE/FreeTube/pull/62 into this PR

Will copy stuff from it but to see the diff please visit that PR

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Everything LGTM but i will leave it to 3 others to give the approval because i helped tested this feature since the beginning so i could overlook bunch of stuff

jasonhenriquez commented:
@absidue @ChunkyProgrammer

PikachuEXE commented:
Just resolve conflict with https://github.com/FreeTubeApp/FreeTube/pull/4376 on thumbnail style (block for `.thumbnailImage` got changed, and to resolve the conflict I brought the old block back but without the `inline-size`)

Thumbnail (especially playlist thumbnail with placeholder image when empty) should still look fine

![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/edce4aec-a7ce-4eed-9ada-762a305fc70d)
![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/f2ab0788-0136-49ca-a486-d1073bb88f42)

PikachuEXE reviewed (commented):
From @absidue 

> Not a fan of unconditionally delaying the loading of components like the watch page until local playlists are loaded, even when you aren't watching a video with any kind of playlist. 
Also the current implementation checks if the id is a local playlist by checking if it exists in the local playlists store. 
That could be done with a simple url parameter (playlistType=user, default if the url parameter wasn't passed is that it is a remote playlist), that way you would only have to delay the loading if the url parameter says it's a local playlist, instead of also delaying when it's remote playlist.

> Basically you can only get to a watch page with a local playlist, if you came from a page that already knows it's a local playlist, so you can just pass that information on in a url parameter, instead of having to detect it from scratch again. 
That also solves the problem of local playlist's having an id that potentially matches a remote one. 

PikachuEXE commented:
Updated Watch view to only delay loading when user playlist requested
Also fixed single playlist view cover image link not passing playlist type for user playlist

Tested with
- video with no playlist
- video with remote playlist
- video with user playlist

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Stock thumbnail should not be clickable because it cant take you to the first vid in the list.

https://github.com/FreeTubeApp/FreeTube/assets/73130443/05365669-19dc-40c9-9923-350bbb5817cd

efb4f5ff-1298-471a-8973-3d47447115dc commented:
If i hit enter instead of the create button im able to create mutiple playlists

https://github.com/FreeTubeApp/FreeTube/assets/73130443/c2921885-31b7-4b6a-9e4d-d7db86df7791

PikachuEXE commented:
@efb4f5ff-1298-471a-8973-3d47447115dc

I copied https://youtube.com/playlist?list=PLm4EwY0oINs9jnQg_xvMz1nfJJ_8TAfcS to a local playlist **twice** and see no such warning in that playlist
Please provide steps to reproduce

Will address other stuff later

PikachuEXE commented:
> If i hit enter instead of the create button im able to create multiple playlists

The `Create New Playlist` button receive a click event (`PointerEvent` to be precise) when enter is pressed on input box in create new playlist prompt input box
No idea how to fix this
![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/97de3301-e4f5-414e-8780-ce49215fa06a)

PikachuEXE commented:
> Clear input field icon isnt centered anymore

I think one of those input tag related change might have done something
Anyway fixed

![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/b4a6aa6a-3f02-4baa-a77e-bd5807fcf50e)

PikachuEXE commented:
> If i randomly click somewhere in the add to playlist window and press enter it closes the window

I think this is intentional
The code is from old branch so if you guys think this behaviour is unwanted I can remove it easily

![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/d514e9df-dffe-4346-8021-e16baa56d81f)
![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/897061bd-d686-4e5f-96fd-d06fe8d31a6c)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> Got this vue warning when opening up my playlist. I think this happened because i copied the remote playlist twice into the local one



https://github.com/FreeTubeApp/FreeTube/assets/73130443/41e3f767-e2d9-4666-814a-736cf75c245f



> Stock thumbnail should not be clickable because it cant take you to the first vid in the list.

Fixed

> If i hit enter instead of the create button im able to create multiple playlists

It still allows the user to do this a few times when the user doesnt have playlists 



https://github.com/FreeTubeApp/FreeTube/assets/73130443/df4dad6c-902a-44e8-bffc-d09c2d32f78e



> This button button color fades in and out kind of weirdly

Its not a bug that the button changes color from blue to red and red to blue but the bug is that u see the transition of the button color
 
> Clear input field icon isnt centered anymore

Also here not centered 

![Capture](https://github.com/FreeTubeApp/FreeTube/assets/73130443/be24fc10-e661-4a5f-a47e-63c54f20af16)

New bug: Add a video from channel page video tab to your playlist -> You will see the description of the video in your playlist. Add the same video to your playlist but this time the video originates from a playlist -> You will see that there is no description of the video. Expected behavior is that the description is always hidden.

![Capture](https://github.com/FreeTubeApp/FreeTube/assets/73130443/b1537757-767b-4d7e-9f5e-91e612b584b7)

PikachuEXE commented:
> Got this vue warning when opening up my playlist. I think this happened because i copied the remote playlist twice into the local one

Got a fix but might not be error proof fix (but solves your case)

PikachuEXE commented:
> It still allows the user to do this a few times when the user doesnt have playlists

No idea how to fix this help

PikachuEXE commented:
> It still allows the user to do this a few times when the user doesnt have playlists

To workaround this issue, I have changed the search input box on add video to playlist prompt to be always visible
Not a good UX but still better than having that issue

> This button button color fades in and out kind of weirdly

Please test on existing `ft-icon-button` with same theme to see if it happens too, I don't think a new button theme/style is introduced here

PikachuEXE commented:
> New bug: Add a video from channel page video tab to your playlist -> You will see the description of the video in your playlist. Add the same video to your playlist but this time the video originates from a playlist -> You will see that there is no description of the video. Expected behavior is that the description is always hidden.

Fixed, along with view count

> you can navigate to the deleted playlist

Replaced with toast message + loading screen
![image](https://github.com/FreeTubeApp/FreeTube/assets/1018543/aeb4b405-7f00-4930-9ebd-f3f7812d46eb)

> Also here not centered

Fixed

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> Got this vue warning when opening up my playlist. I think this happened because i copied the remote playlist twice into the local one 

Still getting this one with the same steps as seen in the video

>  If i hit enter instead of the create button im able to create multiple playlists
>>It still allows the user to do this a few times when the user doesnt have playlists

Fixed



> Clear input field icon isnt centered anymore
>>Also here not centered

Fixed

> New bug: Add a video from channel page video tab to your playlist -> You will see the description of the video in your playlist. Add the same video to your playlist but this time the video originates from a playlist -> You will see that there is no description of the video. Expected behavior is that the description is always hidden.

Fixed

>    you can navigate to the deleted playlist
>> Replaced with toast message + loading screen

Im still seeing the behavior of the video

**New Bug:** search suggestions blocked by thumbnail

![Capture](https://github.com/FreeTubeApp/FreeTube/assets/73130443/501af059-eb5a-4d7b-8543-eb9378b1b0dd)

PikachuEXE commented:
> Still getting this one with the same steps as seen in the video

Real fix applied (I know why it happens now see commit message for details

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> Got this vue warning when opening up my playlist. I think this happened because i copied the remote playlist twice into the local one 
>> Still getting this one with the same steps as seen in the video

> you can navigate to the deleted playlist
>> Replaced with toast message + loading screen
>>>Im still seeing the behavior of the video

> New Bug: search suggestions blocked by thumbnail

All fixed. 

Not sure if the loading screen is the way to go in terms of providing a good UX. Maybe a blank page would be better but i would like the others to chime into that.

New Bug: Reverse playlist throws error


https://github.com/FreeTubeApp/FreeTube/assets/73130443/e2ca4d3b-e8ce-456a-a3c0-60c42590a3ad

PikachuEXE commented:
404 view for user playlist can be updated in later PRs (it's an edge case)

> Reverse playlist throws error

Fixed by creating a new array due to `this.playlistItems` can be arrays from data store state
Which affects **both** cached playlist (local API I guess) and user playlist
[`Array#reverse`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reverse) does mutation in place (there is also [`toReversed`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toReversed) but `reverse` browser compatibility is much better)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
All seems fine now!

PikachuEXE commented:
Anything else to be fixed/changed?

Just merged `development` and [started a build](https://github.com/PikachuEXE/FreeTube/actions/runs/7148274043) in case those can't compile want to test 
(or [look for latest build for this branch as a new one might be created later](https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Afeature%2Fplaylist-2023-05))
**Always** use non official releases with backup, or even the official releases

absidue commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.css:
```diff
@@ -0,0 +1,28 @@
+    along with FreeTube.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+* Credit goes to pavelvaravko for making this css.
+* https://codepen.io/pavelvaravko/pen/qjojOr
+*/
```
Is this credit message still relevant?

PikachuEXE replied:
It's from old code
I guess it's some copy pasta
Probably can be removed

senntore commented:
@PikachuEXE I tested coping playlists with huge number of videos(something like 50+ & 100+) & when I click this
> Some videos in the playlist are not loaded yet. Click here to copy anyway.

almost all of the time, I was able to copy up to 100 videos in a playlist without trying to load more videos in the playlist by scrolling down. So it seems we don't need this message for playlist with less than 100 videos.
Though copying playlist with 100+ videos can be a pain as I had to scroll all the way to their last video, so that all videos can be copied.

absidue commented:
@senntore see this comment further up in the thread and the ones surrounding it: https://github.com/FreeTubeApp/FreeTube/pull/4234#issuecomment-1793199615

PikachuEXE commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.css:
```diff
@@ -0,0 +1,50 @@
+/*
+    This file is part of FreeTube.
+    FreeTube is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    FreeTube is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with FreeTube.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
```
```suggestion
```

PikachuEXE commented on src/renderer/components/ft-create-playlist-prompt/ft-create-playlist-prompt.css:
```diff
@@ -0,0 +1,28 @@
+/*
+    This file is part of FreeTube.
+    FreeTube is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    FreeTube is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with FreeTube.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+* Credit goes to pavelvaravko for making this css.
+* https://codepen.io/pavelvaravko/pen/qjojOr
+*/
+
+/* select starting stylings ------------------------------*/
```
```suggestion
```

PikachuEXE reviewed (commented):
(Comment for remove comments that are absent in other current CSS files)

PikachuEXE commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.scss:
```diff
@@ -0,0 +1,280 @@
+    position: relative;
+
+    .optionsButton {
+      float: right; // ohhhh man, float was finally the right choice for something;
```
```suggestion
      float: right;
```

jasonhenriquez commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.scss:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    .externalPlayerButton {
+      float: right;
```
```suggestion
      float: var(--float-right-ltr-rtl-value);
```

jasonhenriquez commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.scss:
```diff
@@ -0,0 +1,280 @@
+    position: relative;
+
+    .optionsButton {
+      float: right;
```
```suggestion
      float: var(--float-right-ltr-rtl-value);
```

PikachuEXE commented:
Fixed float (`yarn lint-fix` won't detect it

PikachuEXE commented:
Merged https://github.com/FreeTubeApp/FreeTube/pull/4437 and revert fixed in this branch
Clear text button style should still be fine

jasonhenriquez commented on src/renderer/components/ft-playlist-selector/ft-playlist-selector.scss:
```diff
@@ -0,0 +1,280 @@
+$thumbnail-overlay-opacity: 0.85;
+$watched-transition-duration: 0.5s;
```
**issue (non-blocking):** Many of the styling rules in this file seem to not actually be matching anything. I would say it's copied from somewhere, but the styling seems to be unique in some parts for reasons I can't discern (e.g., the `videoDuration` class rules).

jasonhenriquez replied:
If at least some of this is important but is just rote copying, my suggestion would be to create a `mixin` for that and just import it here.

PikachuEXE replied:
It's too early to call for style variable extraction
Since after removing unused style they become unused

absidue commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.vue:
```diff
@@ -0,0 +1,70 @@
+      ref="searchBar"
+      :placeholder="$t('User Playlists.AddVideoPrompt.Search in Playlists')"
+      :show-clear-text-button="true"
+      :clear-input-text-button-vertically-centered="true"
```
```suggestion
```

PikachuEXE replied:
Done

efb4f5ff-1298-471a-8973-3d47447115dc commented:
I do cant find any weird bug anymore. I do have a question though about the sorting of playlists.

When the user is on the playlist page, the default sorting state is Recently played. 
User changes the sorting to for example A-Z -> moves away from the playlist page -> navigate back to it -> still sorted by A-Z (which is good imo) -> user closes FT completely -> opens FT -> navigate to playlist page -> sorting is back to default state, Recently played

Q: Would it be better to always remember the last sorted state instead of default, so in this case that would mean the user gets to see sorted by A-Z when they open FT

Add to playlist prompt behaves differently as it will always default to Recently played even when you changed it

Same Q basically applies here, should we remember last set sorting state?

PikachuEXE commented:
I think people change sorting order to find specific playlist without filtering
But they don't find the same playlist every time
Which is why I default sorting to `latest played first` (playing last played playlist > find playlist in terms of frequency)

For add to playlist prompt let me take a look

PikachuEXE commented:
> Add to playlist prompt behaves differently as it will always default to Recently played even when you changed it

I doubt it when I read it (coz I code it)
It is `recently updated first` (always)

It's more common IMO to add to the same playlist, so setting `recently updated first` would be beneficial for that.

jasonhenriquez commented on src/renderer/components/data-settings/data-settings.js:
```diff
@@ -882,87 +886,116 @@ export default defineComponent({
             })
 
             playlistObject[key] = videoArray
+          } else if (ignoredKeys.includes(key)) {
+            // Do nothing for keys to be ignored
           } else {
```
```suggestion
          } else if (!ignoredKeys.includes(key)) {
            // Do nothing for keys to be ignored
```

<sub>**Obligatory disclaimer:** I am prone to lexical errors when writing in the GitHub UI</sub>

PikachuEXE replied:
https://github.com/FreeTubeApp/FreeTube/pull/4234/commits/36e7ed883afc535d0004df0445a46da2e780ab94

jasonhenriquez commented:
**thoughts:** Thinking about UX considerations with copying an existing playlist. These are non-blocking and out of scope for this PR, but it is worth discussing if it can help inform our future direction.

i. This PR has a button to copy the videos in a YT playlist to a user playlist. This behavior will be possible through #4198 (Enable Video Selection Mode => Ctrl+A => Add to Playlist). Once we have that, it's worth considering if we want its own button for that, or if we want to repurpose it somehow.

ii. Continuing from i), I feel like "Saving a playlist" meaning "copy all of its videos to a user playlist" is not what a lot of people want, most of the time. To compare to YT, saving an existing YTer's playlist pretty much just adds a link to that playlist page on your sidebar. This is a pretty reasonable choice. I don't regularly want to do the equivalent of "forking" my YTer's playlist like we are allowing; rather, I want to "subscribe" to it to see what videos they've updated it with, and be able to access it with ease.

I think we at FreeTube could benefit from doing something similar here; basically just have a "Save Playlist" button that saves the link, and puts the playlist in my Playlists section. I would argue that making it immutable (as in, no removing or adding of videos, and they don't show up in the "Add to Playlist" modal), and non-renameable, are also worth doing to make this "link"/"subscription" relationship for these types of playlists even clearer to the user. It may also help to reduce the extra requests & UX annoyance of having to keep scrolling & loading more videos to "fully" copy a playlist, which is currently pretty clunky.

PikachuEXE commented:
You are describing subscribing/bookmarking a **remote** playlist which is totally different from managing **local** playlists.
I also don't understand "Saving a playlist" (ii) point when the button on remote playlist is labelled "Copy Playlist"
![2023-12-16 15_53_44-Playlist - FreeTube](https://github.com/FreeTubeApp/FreeTube/assets/1018543/d041d6c3-8404-421a-b358-c1039bf05506)

This PR is about **local** playlist only.

jasonhenriquez commented on src/renderer/components/ft-playlist-add-video-prompt/ft-playlist-add-video-prompt.js:
```diff
@@ -0,0 +1,258 @@
+        this.updatePlaylist({ _id: playlist._id })
+      })
+
+      showToast(this.$t('User Playlists.AddVideoPrompt.Toast["{videoCount} video(s) added to {playlistCount} playlist(s)."]', {
```
**suggestion (non-blocking):** use `tc` and update label with pluralization in lieu of "(s)".

PikachuEXE replied:
https://github.com/FreeTubeApp/FreeTube/pull/4234/commits/e7729eecc4c88a28c0385b267ac233603eeba006

There are 2 counts so 2 translation entries are used

jasonhenriquez commented on src/renderer/components/playlist-info/playlist-info.js:
```diff
@@ -79,49 +168,160 @@ export default defineComponent({
+        }
+        try {
+          this.updatePlaylist(playlist)
+          showToast(this.$t('User Playlists.SinglePlaylistView.Toast["{videoCount} video(s) have been removed."]', {
```
**suggestion (non-blocking):** use `tc` and update label with pluralization in lieu of "(s)".

PikachuEXE replied:
https://github.com/FreeTubeApp/FreeTube/pull/4234/commits/e7729eecc4c88a28c0385b267ac233603eeba006

PikachuEXE commented:
Download latest build if you want to
https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Afeature%2Fplaylist-2023-05

PikachuEXE commented:
Just fixed a bug found when I use the build myself: playlist link in watch page not having `playlistType`

I did a search in code and there should be no other places having link to a local user playlist

[`6647099` (#4234)](https://github.com/FreeTubeApp/FreeTube/pull/4234/commits/6647099ba099093c7824dfdc9e8dbeb4b96c8de8)

JohnBetaro commented:
What about the ability to go to any channel and save a whole playlist from there directly? I think this should be a necessary feature' any idea if it will be implemented? Thank you for your time.

absidue commented:
The "Remove Watched Videos" button is displayed inside of playlists, even when "Remember History" is disabled, is that intentional?

efb4f5ff-1298-471a-8973-3d47447115dc commented:
I think it is, disabling that feature applies only on future watched videos not previously watched videos
